File: packages/sdk-python/llm_orchestra/cache/semantic_cache.py
from __future__ import annotations

import copy
import inspect
import math
import time
from dataclasses import dataclass
from typing import Any, Dict, List, Optional

from ..types import (
    CacheConfig,
    CacheEmbeddingFunction,
    CacheEmbeddingInput,
    CacheKeyFunction,
    CompletionRequest,
    CompletionResponse,
)

DEFAULT_SIMILARITY_THRESHOLD = 0.9
DEFAULT_TTL_SECONDS = 3600
DEFAULT_MAX_SIZE = 500
DEFAULT_EMBEDDING_DIM = 128


@dataclass
class CacheEntry:
    signature: str
    embedding: List[float]
    response: CompletionResponse
    created_at: float
    last_accessed: float


class SemanticCache:
    def __init__(self, config: CacheConfig) -> None:
        self.config = {
            "ttlSeconds": config.get("ttlSeconds", DEFAULT_TTL_SECONDS),
            "maxSize": config.get("maxSize", DEFAULT_MAX_SIZE),
            "similarityThreshold": config.get("similarityThreshold", DEFAULT_SIMILARITY_THRESHOLD),
        }
        self.embedder: CacheEmbeddingFunction = config.get("embeddingFunction") or default_embedder
        self.key_function: CacheKeyFunction = config.get("keyFunction") or default_key_function
        self.entries: List[CacheEntry] = []

    async def get(self, request: CompletionRequest) -> Optional[CompletionResponse]:
        signature = self.key_function(request)
        self._prune_expired()

        candidates = [entry for entry in self.entries if entry.signature == signature]
        if not candidates:
            return None

        embedding = await _maybe_await(
            self.embedder({"text": build_embedding_text(request), "request": request})
        )

        best_match: Optional[CacheEntry] = None
        best_score = self.config["similarityThreshold"]

        for entry in candidates:
            score = cosine_similarity(embedding, entry.embedding)
            if score >= best_score:
                best_score = score
                best_match = entry

        if not best_match:
            return None

        best_match.last_accessed = time.time()
        return copy.deepcopy(best_match.response)

    async def set(self, request: CompletionRequest, response: CompletionResponse) -> None:
        signature = self.key_function(request)
        embedding = await _maybe_await(
            self.embedder({"text": build_embedding_text(request), "request": request})
        )

        now = time.time()
        entry = CacheEntry(
            signature=signature,
            embedding=embedding,
            response=copy.deepcopy(response),
            created_at=now,
            last_accessed=now,
        )

        self.entries.append(entry)
        self._prune_expired()
        self._prune_overflow()

    def clear(self) -> None:
        self.entries = []

    def _prune_expired(self) -> None:
        now = time.time()
        max_age = self.config["ttlSeconds"]
        self.entries = [
            entry for entry in self.entries if (now - entry.created_at) <= max_age
        ]

    def _prune_overflow(self) -> None:
        max_size = self.config["maxSize"]
        if len(self.entries) <= max_size:
            return
        self.entries.sort(key=lambda entry: entry.last_accessed, reverse=True)
        self.entries = self.entries[:max_size]


async def _maybe_await(value):
    if inspect.isawaitable(value):
        return await value
    return value


def default_key_function(request: CompletionRequest) -> str:
    return stable_stringify(
        {
            "model": request.get("model"),
            "temperature": request.get("temperature"),
            "maxTokens": request.get("maxTokens"),
            "topP": request.get("topP"),
            "stop": request.get("stop"),
            "toolChoice": request.get("toolChoice"),
            "tools": request.get("tools"),
        }
    )


async def default_embedder(input_data: CacheEmbeddingInput) -> List[float]:
    tokens = tokenize(input_data.get("text", ""))
    if not tokens:
        return [0.0] * DEFAULT_EMBEDDING_DIM

    vector = [0.0] * DEFAULT_EMBEDDING_DIM
    for token in tokens:
        index = hash_token(token) % DEFAULT_EMBEDDING_DIM
        vector[index] += 1.0

    return normalize_vector(vector)


def build_embedding_text(request: CompletionRequest) -> str:
    segments: List[str] = []
    for message in request.get("messages", []):
        role = message.get("role", "")
        name = f":{message.get('name')}" if message.get("name") else ""
        tool_id = f":{message.get('toolCallId')}" if message.get("toolCallId") else ""
        content = message.get("content", "")
        segments.append(f"{role}{name}{tool_id}:{content}")
    return "\n".join(segments)


def cosine_similarity(a: List[float], b: List[float]) -> float:
    if len(a) != len(b) or not a:
        return 0.0
    dot = sum(x * y for x, y in zip(a, b))
    norm_a = math.sqrt(sum(x * x for x in a))
    norm_b = math.sqrt(sum(x * x for x in b))
    if norm_a == 0 or norm_b == 0:
        return 0.0
    return dot / (norm_a * norm_b)


def normalize_vector(vector: List[float]) -> List[float]:
    norm = math.sqrt(sum(x * x for x in vector))
    if norm == 0:
        return vector
    return [x / norm for x in vector]


def tokenize(text: str) -> List[str]:
    return [token for token in text.lower().split() if token]


def hash_token(token: str) -> int:
    hash_value = 2166136261
    for char in token:
        hash_value ^= ord(char)
        hash_value = (hash_value * 16777619) & 0xFFFFFFFF
    return hash_value


def stable_stringify(value: Any) -> str:
    if value is None or isinstance(value, (str, int, float, bool)):
        return repr(value)
    if isinstance(value, list):
        return "[" + ",".join(stable_stringify(item) for item in value) + "]"
    if isinstance(value, dict):
        items = sorted(value.items(), key=lambda item: item[0])
        return "{" + ",".join(f"{repr(k)}:{stable_stringify(v)}" for k, v in items) + "}"
    return repr(value)

File: src/cache/semantic-cache.ts
/**
 * Semantic cache for LLM responses
 */

import type {
  CompletionRequest,
  CompletionResponse,
  CacheConfig,
  CacheEmbeddingFunction,
  CacheKeyFunction,
  CacheEmbeddingInput,
} from '../types/index.js';

interface CacheEntry {
  signature: string;
  embedding: number[];
  response: CompletionResponse;
  createdAt: number;
  lastAccessed: number;
}

const DEFAULT_SIMILARITY_THRESHOLD = 0.9;
const DEFAULT_TTL_SECONDS = 3600;
const DEFAULT_MAX_SIZE = 500;
const DEFAULT_EMBEDDING_DIM = 128;

export class SemanticCache {
  private entries: CacheEntry[] = [];
  private config: Required<Pick<CacheConfig, 'ttlSeconds' | 'maxSize' | 'similarityThreshold'>>;
  private embedder: CacheEmbeddingFunction;
  private keyFunction: CacheKeyFunction;

  constructor(config: CacheConfig) {
    this.config = {
      ttlSeconds: config.ttlSeconds ?? DEFAULT_TTL_SECONDS,
      maxSize: config.maxSize ?? DEFAULT_MAX_SIZE,
      similarityThreshold: config.similarityThreshold ?? DEFAULT_SIMILARITY_THRESHOLD,
    };
    this.embedder = config.embeddingFunction ?? defaultEmbedder;
    this.keyFunction = config.keyFunction ?? defaultKeyFunction;
  }

  async get(request: CompletionRequest): Promise<CompletionResponse | null> {
    const signature = this.keyFunction(request);
    this.pruneExpired();

    const candidates = this.entries.filter((entry) => entry.signature === signature);
    if (candidates.length === 0) return null;

    const embedding = await this.embedder({
      text: buildEmbeddingText(request),
      request,
    });

    let bestMatch: CacheEntry | null = null;
    let bestScore = this.config.similarityThreshold;

    for (const entry of candidates) {
      const score = cosineSimilarity(embedding, entry.embedding);
      if (score >= bestScore) {
        bestScore = score;
        bestMatch = entry;
      }
    }

    if (!bestMatch) return null;

    bestMatch.lastAccessed = Date.now();
    return cloneResponse(bestMatch.response);
  }

  async set(
    request: CompletionRequest,
    response: CompletionResponse
  ): Promise<void> {
    const signature = this.keyFunction(request);
    const embedding = await this.embedder({
      text: buildEmbeddingText(request),
      request,
    });

    const entry: CacheEntry = {
      signature,
      embedding,
      response: cloneResponse(response),
      createdAt: Date.now(),
      lastAccessed: Date.now(),
    };

    this.entries.push(entry);
    this.pruneExpired();
    this.pruneOverflow();
  }

  clear(): void {
    this.entries = [];
  }

  private pruneExpired(): void {
    const now = Date.now();
    const maxAgeMs = this.config.ttlSeconds * 1000;
    this.entries = this.entries.filter((entry) => now - entry.createdAt <= maxAgeMs);
  }

  private pruneOverflow(): void {
    if (this.entries.length <= this.config.maxSize) return;
    this.entries.sort((a, b) => b.lastAccessed - a.lastAccessed);
    this.entries = this.entries.slice(0, this.config.maxSize);
  }
}

function defaultKeyFunction(request: CompletionRequest): string {
  return stableStringify({
    model: request.model,
    temperature: request.temperature ?? null,
    maxTokens: request.maxTokens ?? null,
    topP: request.topP ?? null,
    stop: request.stop ?? null,
    toolChoice: request.toolChoice ?? null,
    tools: request.tools ?? null,
  });
}

const defaultEmbedder: CacheEmbeddingFunction = async (
  input: CacheEmbeddingInput
): Promise<number[]> => {
  const tokens = tokenize(input.text);
  if (tokens.length === 0) {
    return new Array(DEFAULT_EMBEDDING_DIM).fill(0);
  }

  const vector = new Array(DEFAULT_EMBEDDING_DIM).fill(0);
  for (const token of tokens) {
    const index = hashToken(token) % DEFAULT_EMBEDDING_DIM;
    vector[index] += 1;
  }

  return normalizeVector(vector);
};

function buildEmbeddingText(request: CompletionRequest): string {
  return request.messages
    .map((message) => {
      const name = message.name ? `:${message.name}` : '';
      const toolId = message.toolCallId ? `:${message.toolCallId}` : '';
      return `${message.role}${name}${toolId}:${message.content}`;
    })
    .join('\n');
}

function cosineSimilarity(a: number[], b: number[]): number {
  if (a.length !== b.length || a.length === 0) return 0;
  let dot = 0;
  let normA = 0;
  let normB = 0;
  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i];
    normA += a[i] * a[i];
    normB += b[i] * b[i];
  }
  if (normA === 0 || normB === 0) return 0;
  return dot / (Math.sqrt(normA) * Math.sqrt(normB));
}

function normalizeVector(vector: number[]): number[] {
  const norm = Math.sqrt(vector.reduce((sum, value) => sum + value * value, 0));
  if (norm === 0) return vector;
  return vector.map((value) => value / norm);
}

function tokenize(text: string): string[] {
  return text
    .toLowerCase()
    .split(/\s+/)
    .map((token) => token.trim())
    .filter(Boolean);
}

function hashToken(token: string): number {
  let hash = 2166136261;
  for (let i = 0; i < token.length; i++) {
    hash ^= token.charCodeAt(i);
    hash = Math.imul(hash, 16777619);
  }
  return hash >>> 0;
}

function stableStringify(value: unknown): string {
  if (value === null || typeof value !== 'object') {
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return `[${value.map((item) => stableStringify(item)).join(',')}]`;
  }
  const entries = Object.entries(value as Record<string, unknown>)
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([key, val]) => `${JSON.stringify(key)}:${stableStringify(val)}`);
  return `{${entries.join(',')}}`;
}

function cloneResponse(response: CompletionResponse): CompletionResponse {
  if (typeof structuredClone === 'function') {
    return structuredClone(response);
  }
  return JSON.parse(JSON.stringify(response)) as CompletionResponse;
}
